crates/krator-derive/src/admission.rs

```diff
@@ -0,0 +1,273 @@
+                            ..Default::default()
+                        },
+                        string_data: Some(data),
+                        type_: Some("tls".to_string()),
+                        ..Default::default()
+                    }
+                }
+
+                /// Returns a service that forwards to pods where label `app` has the value returned by the function
+                /// `admission_webhook_service_app_selector()`. It exposes port `443` (necessary for webhooks) and
+                /// listens to the pod's port `8443`
+                pub fn admission_webhook_service(namespace: &str) -> k8s_openapi::api::core::v1::Service {
+                    let crd = #name_identifier::crd();
+
+                    let mut selector = std::collections::BTreeMap::new();
+                    selector.insert("app".into(), #name_identifier::admission_webhook_service_app_selector());
+
+                    k8s_openapi::api::core::v1::Service {
+                        metadata: k8s_openapi::apimachinery::pkg::apis::meta::v1::ObjectMeta {
+                            name: Some(#name_identifier::admission_webhook_service_name()),
+                            namespace: Some(namespace.to_string()),
+                            ..Default::default()
+                        },
+                        spec: Some(k8s_openapi::api::core::v1::ServiceSpec {
+                            selector: Some(selector),
+                            ports: Some(vec![k8s_openapi::api::core::v1::ServicePort{
+                                protocol: Some("TCP".to_string()),
+                                port: 443,
+                                target_port: Some(k8s_openapi::apimachinery::pkg::util::intstr::IntOrString::Int(8443)),
+                                ..Default::default()
+                            }]),
+                            type_: Some("ClusterIP".to_string()),
+                            ..Default::default()
+                        }),
+                        status: None
+                   }
+                }
+
+                /// Creates a MutatingWebhookConfiguration using the certificate from the given service and the service
+                /// of the given service as configuration
+                pub fn admission_webhook_configuration(service: k8s_openapi::api::core::v1::Service, secret: k8s_openapi::api::core::v1::Secret) -> anyhow::Result<k8s_openapi::api::admissionregistration::v1::MutatingWebhookConfiguration> {
+                   use anyhow::Context;
+                   let crd = #name_identifier::crd();
+
+                   let webhook_name = #name_identifier::admission_webhook_configuration_name() ;
+                   let versions: Vec<String> = crd.spec.versions.into_iter().map(|v| v.name).collect();
+
+                   anyhow::ensure!(secret.type_ == Some("tls".to_string()), format!("secret with name {} is not a tls secret", secret.metadata.name.unwrap()));
```

Comment:
Are `unwrap`s the correct thing here? Seeing as these are errors, we should probably do an `unwrap_or_default`

Reply:
fixed in commit 5d54163 ... I am catching errors early with `ensure!` to have more speaking error messages